autodocodec/src/Autodocodec/Aeson/Decode.hs

```diff
@@ -77,6 +81,9 @@ parseJSONContextVia codec_ context_ =
           (\object_ -> (`go` c) (object_ :: JSON.Object))
       HashMapCodec c -> liftParseJSON (`go` c) (`go` listCodec c) value :: JSON.Parser (HashMap _ _)
       MapCodec c -> liftParseJSON (`go` c) (`go` listCodec c) value :: JSON.Parser (Map _ _)
+#if MIN_VERSION_aeson(2,0,0)
+      KeyMapCodec c -> liftParseJSON (`go` c) (`go` listCodec c) value :: JSON.Parser (KeyMap _)
```

Comment:
In essence the `KeyMapCodec` will always equal either `HashMapCodec` or `MapCodec`, depending on which value the aeson security flag has.
I can't tell whether we should make `keyMapCodec` a function that choses the right one based on that flag instead of making an extra constructor.

Reply:
It feels like the best way to make this package not care about the internal implementation details of aeson (ie. whether `aeson`'s `KeyMap` is a `HashMap` or a `Map`) would be to have the `KeyMapCodec` constructor. Is it even possible to access the value of aeson's `ordered-keymap` flag in this package in order to switch `keyMapCodec` implementation?

The problem with the KeyMapCodec, however, is that because we're using CPP conditionals here to make the package work for both aeson 1 and 2, this means that the shape of the Codec type is different depending on whether aeson 1 or 2 is referenced (ie. sometimes it does not have the new constructor). To have a breaking change like that dependent on the version of a referenced package makes me quite uncomfortable. For example, I imagine the Hackage doco would include the new constructor, but this would not be true when compiling for aeson 1.

I wonder if it would not be better to just ditch aeson 1 support on master and just support aeson 2, and do a major version increment. We could have a branch for bugfixes against the aeson 1 version, so as to not completely abandon those people using older aeson. But going forward it feels like the ecosystem is going to more and more be built on top of aeson 2 and care less and less about aeson 1. Throwing away aeson 1 support on master would also mean we could get rid of the compat code and all the CPP too. 

If we could just flip implementations transparently with CPP and not break our types, then I'd be more in favour of supporting both aeson 1 and 2 on master simultaneously... thoughts?

Reply:
Okay, so I just saw how @NorfairKing achieved making keyMapCodec function that either uses MapCodec or HashMapCodec ([here](https://github.com/NorfairKing/autodocodec/blob/de3a6efbfad165d2265f9d222a570d9a938c162d/autodocodec/src/Autodocodec/Codec.hs#L973)). 

I think I like this approach better than introducing a new `KeyMapCodec` constructor. It means the Codec type pretty much stays the same whether aeson 1 or 2 is used. It's a shame that this is even necessary, but if we need to maintain support for both aeson 1 and 2, that seems like the least disruptive approach.